sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -863,7 +863,50 @@ protected BigtableSource withEstimatedSizeBytes(Long estimatedSizeBytes) {
           Math.max(sizeEstimate / maximumNumberOfSplits, desiredBundleSizeBytes);
 
       // Delegate to testable helper.
-      return splitBasedOnSamples(desiredBundleSizeBytes, getSampleRowKeys(options));
+      List<BigtableSource> splits =
+          splitBasedOnSamples(desiredBundleSizeBytes, getSampleRowKeys(options));
+      return reduceSplits(splits, options);
+    }
+
+    private List<BigtableSource>
+        reduceSplits(List<BigtableSource> splits, PipelineOptions options) throws IOException {
+      final long maximumCountOfSplits = 15_360L;
+      if (splits.size() < maximumCountOfSplits) {
+        return splits;
+      }
+      ImmutableList.Builder<BigtableSource> reducedSplits = ImmutableList.builder();
+      BigtableSource start = null;
+      BigtableSource lastSeen = null;
+      int numberToCombine =
+          (int) ((splits.size() + maximumCountOfSplits - 1) / maximumCountOfSplits);
+      int counter = 0;
+      long size = 0;
+      for (BigtableSource source : splits) {
+        if (counter == 0) {
+          start = source;
+        }
+        size += source.getEstimatedSizeBytes(options);
+        counter++;
+        lastSeen = source;
+        if (counter == numberToCombine) {
+          ByteKey startKey = start.getRanges().get(0).getStartKey();
+          ByteKey endKey = source.getRanges().get(source.getRanges().size() - 1).getEndKey();
+          reducedSplits.add(
+              start.withSingleRange(ByteKeyRange.of(startKey, endKey))
```

Comment:
@chamikaramj, good catch.  It's a correctness issue.  Would it make sense to simply add the two sets of ranges together?

Reply:
Thanks @sduskis and @chamikaramj  for looking into this, I can rewrite the function as below:

    private List<BigtableSource>
        reduceSplits(List<BigtableSource> splits, PipelineOptions options) throws IOException {
      final long maximumCountOfSplits = 15_360L;
      int numberToCombine =
          (int) ((splits.size() + maximumCountOfSplits - 1) / maximumCountOfSplits);
      if (splits.size() < maximumCountOfSplits || numberToCombine < 2) {
        return splits;
      }
      ImmutableList.Builder<BigtableSource> reducedSplits = ImmutableList.builder();
      ImmutableList.Builder<ByteKeyRange> currentSplits = ImmutableList.builder();
      int counter = 0;
      long size = 0;
      for (BigtableSource source : splits) {
        size += source.getEstimatedSizeBytes(options);
        counter++;
        currentSplits.addAll(source.getRanges());
        if (counter == numberToCombine) {
          reducedSplits.add(
              new BigtableSource(config, filter, currentSplits.build(), size));
          counter = 0;
          size = 0;
          currentSplits = ImmutableList.builder();
        }
      }
      if (size > 0) {
        reducedSplits.add(
            new BigtableSource(config, filter, currentSplits.build(), size));
      }
      return reducedSplits.build();
    }

However BigtableReader will run into issues as it works only with single range.

 public BigtableReader(BigtableSource source, BigtableService service) {
      **checkArgument(source.getRanges().size() == 1, "source must have exactly one key range");**
      this.source = source;
      this.service = service;
      rangeTracker = ByteKeyRangeTracker.of(source.getRanges().get(0));
    }

 is there any other way of combining ranges?

Reply:
This is a problem in master now, caused by #4312.  This can never work with multiple key ranges, which was the point of #4312.

`checkArgument(source.getRanges().size() == 1, "source must have exactly one key range");`

@rniemo-g, how does `BigtableReader` work with multiple ranges now?

Reply:
BigtableReader doesn't work with multiple ranges, but BigtableIO.Read does. I.e. BigtableIO.Read takes in multiple key ranges and creates a BigtableSource with those ranges. split() is then called on this multi-range source, such that the output sources are single range. Then a BigtableReader is constructed for each of these sources (AFIAU). That is, the complexity for multiple ranges is mostly in the split() function. (This should admittedly be made more clear with some comments)

I did have BigtableReader support multiple key ranges by having a multi-range version of ByteKeyRangeTracker, and a modified splitAtFraction(), but as Eugene pointed out on #4312 it was unnecessary complexity. That complexity might be needed here?

I also want to point out Igor's comments near the middle on #4312, where we discuss this bug of having too many splits. He suggests moving some of the split() functionality to the Bigtable client, which might be worth looking into as well.

Quick comment on this part of the PR itself:
Combining sources by constructing a keyrange with a start key of one source's range and an end key of another source's range could be dangerous if the orginal source's ranges aren't adjacent. E.g. range [0, 100) of source1 and range [200, 300) of source2 are combined into [0, 300). The range [100, 200) is now read in the combined source, when the original sources did not read this range.

Reply:
Thanks for explaining!

This PR need to check for adjacency of combined sources before merging sources.  that makes sense.

arkash@, let's please discuss this offline.

Reply:
Thanks @rniemo-g. I will follow up with @sduskis.

Reply:
np!

Reply:
@sduskis please review have added a function to check adjacency of ranges before merging.